src/scenic/core/object_types.py

```diff
@@ -3,666 +3,1760 @@
+        x, y = zip(*triangle)
+        plt.fill(x, y, "w")
+        plt.plot(x + (x[0],), y + (y[0],), color="k", linewidth=1)
+
+@distributionFunction
+def defaultSideSurface(occupiedSpace, dimension, positive, thresholds):
+    """Extracts a side surface from the occupiedSpace of an object.
+
+    This function is the default implementation for computing a region
+    representing a side surface of an object. This is done by keeping only the
+    faces of the object's ``occupiedSpace`` mesh that have normal 
+    vectors with a large/small enough x,y, or z component. For example, 
+    for the front surface of an object we would would keep all faces that 
+    had a normal vector with y component greater than ``thresholds[1][1]`` 
+    and for the back surface of an object we would keep all faces that
+    had a normal vector with y component less than ``thresholds[1][0]``.
+
+    Args:
+        occupiedSpace: The `occupiedSpace` region of the object to
+          extract the side surface from.
+        dimension: The target dimension who's component will be checked.
+        positive: If `False`, the target component must be less than 
+          the first value in the appropriate tuple. If `True`, the
+          component must be greater than the second value in the
+          appropriate tuple.
+        thresholds: A 3-tuple of 2-tuples, one for each dimension (x,y,z),
+          with each tuple containing the thresholds for a positive and
+          non-positive side in each dimension.
+        on_dimension: The on_dimension to be passed to the created surface.
+    """
+    # Extract mesh from object
+    obj_mesh = occupiedSpace.mesh.copy()
+
+    # Extract appropriate thresholds
+    threshold = thresholds[dimension][int(positive)]
+
+    # Drop all faces whose normal vector do not have a sufficiently
+    # large component.
+    face_normal_vals = obj_mesh.face_normals[:, dimension]
+    if positive:
+        face_mask = face_normal_vals >= threshold
+    else:
+        face_mask = face_normal_vals <= threshold
+
+    obj_mesh.faces = obj_mesh.faces[face_mask]
+    obj_mesh.remove_unreferenced_vertices()
+
+    # Check if the resulting surface is empty and return an appropriate region.
+    if not obj_mesh.is_empty:
+        return MeshSurfaceRegion(mesh=obj_mesh, center_mesh=False)
+    else:
+        return EmptyRegion(name="EmptyTopSurface")
+
+def canSee(position, orientation, visibleDistance, viewAngles, rayDensity, \
+           target, occludingObjects, debug=False):
+    """Perform visibility checks on Points, OrientedPoints, or Objects, accounting for occlusion.
+
+    For visibilty of Objects:
+
+    1. Do several quick checks to see if the object is naively visible
+       or not visible:
+
+       * If the object contains its position and its position is visible,
+         the object is visible.
+       * If the viewer is inside the object, the object is visible.
```

Comment:
This heuristic is not sound (imagine nested boxes), but it doesn't seem to actually be implemented below either.

Reply:
Oh, it's the `containsPoint` check. I don't think it works, unfortunately.

Reply:
Hmm good point, I'll drop it.